SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularitySchedulerLock.java

```diff
@@ -12,24 +13,65 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(SingularitySchedulerLock.class);
 
-  private final ReentrantLock lock;
+  private final ReentrantLock stateLock;
+  private final ReentrantLock offersLock;
+  private final ConcurrentHashMap<String, ReentrantLock> requestLocks;
 
   @Inject
   public SingularitySchedulerLock() {
-    this.lock = new ReentrantLock();
+    this.stateLock = new ReentrantLock();
+    this.offersLock = new ReentrantLock();
+    this.requestLocks = new ConcurrentHashMap<>();
   }
 
-  public long lock(String name) {
+  public long lock(String requestId, String name) {
     final long start = System.currentTimeMillis();
-    LOG.info("{} - Locking", name);
+    LOG.trace("{} - Locking {}", name, requestId);
+    ReentrantLock lock = requestLocks.computeIfAbsent(requestId, (r) -> new ReentrantLock());
     lock.lock();
-    LOG.info("{} - Acquired ({})", name, JavaUtils.duration(start));
+    LOG.trace("{} - Acquired lock on {} ({})", name, requestId, JavaUtils.duration(start));
     return System.currentTimeMillis();
   }
 
-  public void unlock(String name, long start) {
-    LOG.info("{} - Unlocking ({})", name, JavaUtils.duration(start));
+  public void unlock(String requestId, String name, long start) {
```

Comment:
Should be safe to make `lock()` and `unlock()` private now, and only allow the rest of the scheduler to interact via `runWithRequestLock()`.